pkg/share/manager/cs3/cs3.go

```diff
@@ -0,0 +1,564 @@
+	GatewayAddr       string `mapstructure:"gateway_addr"`
+	ProviderAddr      string `mapstructure:"provider_addr"`
+	ServiceUserID     string `mapstructure:"service_user_id"`
+	ServiceUserIdp    string `mapstructure:"service_user_idp"`
+	MachineAuthAPIKey string `mapstructure:"machine_auth_apikey"`
+}
+
+// NewDefault returns a new manager instance with default dependencies
+func NewDefault(m map[string]interface{}) (share.Manager, error) {
+	c := &config{}
+	if err := mapstructure.Decode(m, c); err != nil {
+		err = errors.Wrap(err, "error creating a new manager")
+		return nil, err
+	}
+
+	s, err := metadata.NewCS3Storage(c.GatewayAddr, c.ProviderAddr, c.ServiceUserID, c.ServiceUserIdp, c.MachineAuthAPIKey)
+	if err != nil {
+		return nil, err
+	}
+	indexer := indexer.CreateIndexer(s)
+
+	return New(s, indexer)
+}
+
+// New returns a new manager instance
+func New(s Storage, indexer Indexer) (*Manager, error) {
+	return &Manager{
+		storage:     s,
+		indexer:     indexer,
+		initialized: false,
+	}, nil
+}
+
+func (m *Manager) initialize() error {
+	if m.initialized {
+		return nil
+	}
+
+	err := m.storage.Init(context.Background(), "cs3-share-manager-metadata")
+	if err != nil {
+		return err
+	}
+	if err := m.storage.MakeDirIfNotExist(context.Background(), "shares"); err != nil {
+		return err
+	}
+	if err := m.storage.MakeDirIfNotExist(context.Background(), "metadata"); err != nil {
+		return err
+	}
+	err = m.indexer.AddIndex(&collaboration.Share{}, option.IndexByFunc{
+		Name: "OwnerId",
+		Func: indexOwnerFunc,
+	}, "Id.OpaqueId", "shares", "non_unique", nil, true)
+	if err != nil {
+		return err
+	}
+	err = m.indexer.AddIndex(&collaboration.Share{}, option.IndexByFunc{
+		Name: "GranteeId",
+		Func: indexGranteeFunc,
+	}, "Id.OpaqueId", "shares", "non_unique", nil, true)
+	if err != nil {
+		return err
+	}
+	m.initialized = true
+	return nil
+}
+
+// Share creates a new share
+func (m *Manager) Share(ctx context.Context, md *provider.ResourceInfo, g *collaboration.ShareGrant) (*collaboration.Share, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+	user := ctxpkg.ContextMustGetUser(ctx)
+	now := time.Now().UnixNano()
+	ts := &typespb.Timestamp{
+		Seconds: uint64(now / 1000000000),
+		Nanos:   uint32(now % 1000000000),
+	}
+
+	share := &collaboration.Share{
+		Id: &collaboration.ShareId{
+			OpaqueId: uuid.NewString(),
+		},
+		ResourceId:  md.Id,
+		Permissions: g.Permissions,
+		Grantee:     g.Grantee,
+		Owner:       md.Owner,
+		Creator:     user.Id,
+		Ctime:       ts,
+		Mtime:       ts,
+	}
+
+	data, err := json.Marshal(share)
+	if err != nil {
+		return nil, err
+	}
+
+	fn := path.Join("shares", share.Id.OpaqueId)
+	err = m.storage.SimpleUpload(ctx, fn, data)
+	if err != nil {
+		return nil, err
+	}
+
+	metadataPath := path.Join("metadata", share.Id.OpaqueId)
+	err = m.storage.MakeDirIfNotExist(ctx, metadataPath)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = m.indexer.Add(share)
+	if err != nil {
```

Comment:
`err` check is unnecessary. Just `return share, err`

Reply:
But if the share couldn't be persisted in the index we don't want the caller to use the share object as that could lead to inconsistencies. I think it's safer to not give that option in the first place, the share hasn't actually be (fully) created.

Reply:
We return a non-nil error. That should be enough indication that the share is not fully initialized and you shouldn't use it at all. I think it is common understanding in go that you do not use a value if the corresponding error is not nil. 

So in my opinion we can safely assume that the `share` value is not used if `err != nil`

Reply:
You are probably right and it'll never be a problem, my personal preference would still have been to keep it a tad cleaner at the cost of an additional `if`. I've also seen both ways being used in go projects. It's fine though, I'll change it and push a commit in a sec.